Doc/library/__main__.rst

```diff
@@ -1,25 +1,196 @@
+is part of two quite different constructs:
+
+1. The ``__name__ == '__main__'`` statement
+2. The ``__main__.py`` file in Python packages
+
+Each of these mechanisms are related to Python modules; how
+users interact with them and how they interact with each other. See
+section :ref:`tut-modules`.
+
+
+``__name__ == '__main__'``
+---------------------------
+
+``'__main__'`` is the name of the environment where top-level code is run.
+"Top-level code" means when a Python module is initialized from an interactive
+prompt, from standard input, from a file argument, from a :option:`-c`
+argument, or from a :option:`-m` argument, but **not** when it is initialized
+from an import statement.  In any of these situations, the module's
+``__name__`` is set equal to ``'__main__'``.
+
+The only other context in which Python code is run is when it is imported
+through an import statement. In that case, ``__name__`` is set equal to the
+module's name: usually the name of the file without the ``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
+        ...
+
+Idiomatic Usage
+^^^^^^^^^^^^^^^
+
+Putting as few statements as possible in the block below ``if __name___ ==
+'__main__'`` can improve code clarity. Most often, a function named *main*
+encapsulates the program's primary behavior, creating this pattern::
+
+    # echo.py
+
+    import shlex
+    import sys
+
+    def echo(phrase: str):
+       # you can imagine that this dummy wrapper around print might be
+       # different and truly worth re-using in a real program.
+       print(phrase)
+
+    def main():
+        "Echo the sys.argv to standard output"
+        echo(shlex.join(sys.argv))
+
+    if __name__ == '__main__':
+        main()
+
+This has the added benefit of the *echo* function itself being isolated and
+importable elsewhere::
+
+    # elsewhere.py
+
+    import sys
+
+    from echo import echo
+
+    def echo_platform():
+        echo(sys.platform)
+
+
+Packaging Considerations (``console_scripts``)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For detailed documentation on Python packaging, see
+`setuptools. <https://setuptools.readthedocs.io/en/latest/>`__
+
+*main* functions are often used to create command line tools by specifying them
+as entry points for console scripts.  When this is done, pip inserts the
+function call into a template that looks like this::
+
+   #!/path/to/python3
+   # -*- coding: utf-8 -*-
+   import re
+   import sys
+   from package.__main__ import main
+   if __name__ == '__main__':
+       sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
+       sys.exit(main())
```

Comment:
I'm a bit wary of this sample code here.

pip uses launcher executables on Windows, for example, which obviously don't look like this -- but more generally, it should be sufficient to say that the call to main is wrapped in sys.exit, without showing the detailed sample code of what the generated scripts could look like.

I'm speaking with the pip maintainer hat on of "hey, why does pip generate a different script than what's documented in Python docs" would be very confusing for me, and feels like something an issue someone would open at some point if we do this. :)

Reply:
@ambv I know @pradyunsg is still marked as requesting changes, but just to bring you up to speed, this example was removed and replaced by a paraphrase of the behavior:

https://github.com/python/cpython/pull/26883/files#diff-ada229cf79d40b37dc14942330de96f7d48f2019cf0b1c864e8644765a4e3e7eR88-R90